### PR TITLE
More Structured format for parameters and Enhanced FES 

### LINKED ISSUE #7752: [2.0]: Properties of objects aren't being checked correctly by FES
### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Build process
- [ ] Unit testing
- [ ] Internationalization
- [x] Friendly errors
- [ ] Other (specify if possible)

### p5.js version

2.0.0

### Web browser and version

Firefox

### Operating system

MacOS

### Steps:
If you run `npm run docs` and then check `docs/parameterData.json`, if you search for `textToModel`, it currently has these parameters:
```
"textToModel": {
      "overloads": [
        [
          "String",
          "Number",
          "Number",
          "Number",
          "Number",
          "Object?",
          "Number?",
          "Number?"
        ]
      ]
    }
```

This is because we're documenting the properties of the object like this:
```
* @param {String} str The text string to convert into a 3D model.
* @param {Number} x The x-coordinate for the starting position of the text.
* @param {Number} y The y-coordinate for the starting position of the text.
* @param {Number} width Maximum width of the text block (wraps text if exceeded).
* @param {Number} height Maximum height of the text block.
* @param {Object} [options] Configuration options for the 3D text:
* @param {Number} [options.extrude=0] The depth to extrude the text. A value of 0 produces
* flat text; higher values create thicker, 3D models.
* @param {Number} [options.sampleFactor=1] A factor controlling the level of detail for the text contours.
*  Higher values result in smoother curves.
```

It looks like we're mistakenly parsing `options.extrude` and `options.sampleFactor` as separate parameters, and not properties of `options`.

This likely needs a change in `utils/convert.js`.

### PR DESCRIPTION
**Pull request by IIITM-Jay**
Resolves #7752 

#### Overall Description
This PR deals with generating more structured and informative `parameterData.json` in order to get a clear and accurate idea of the various parameters involved in overloads of methods. Additionally, as a part of the enhancement process, this PR also resolves and enhances the ability to check the `types` of the nested as well as positional parameters efficiently at run time, which is also one of the critical aspect of the issue.

#### Approach Used
- **Enhancements made to be able to build more structural format to `parameterData.json`**
  - The parameters of  `FunctionType` node are extracted and preserved for further processing 
  - Traversed more deeper to build more informative structure for supporting nested params
  - Used appropriate regular expression as required at appropriate places
  - Efficient handling of:
    - Array Types
    - Union Types
    - Tuple Types
    - Function Parameters
    - Nested Object Parameters
    - Rest Parameters
    - **_Array of Tuples consisting of nested union and other params_**
  
- **Modified FES to parse and check the types at run time via generated `Zod schema`** 
  - Now, string checking via the presence of (`?`) as `endsWith('?')` is no longer required
  - `Switch- Case` is used to handle mappings from `type` names to `Zod schemas`

#### Additional Code Quality Work
- Used fallback mechanism for safety and to avoid runtime errors
- Explicit handling of types to ease overall maintainability and readability
- Added in-line comments as appropriate

#### Trickiest Portions to handle
- Nested Parameters like `[p5.Color|String|Number|Number[], Number][]`
- Rest Parametrs `"...Number[]"`
- Functions with parameters `function(p5.Geometry)`

#### Observations and Results
- Tested and validated the `parameterData.json` to verify the expected results 

#### Additional Checks 
- [x] `npm run docs` - to generate the `parameterData.json` doc
- [x] `npm run lint` - lint checker for code style and standards
- [x] `npm test` - to re-validate all the test suites passed

### CODE REVIEW

File: utils/convert.mjs [hunk 1]
```diff
@@ -512,51 +545,166 @@ function cleanUpClassItems(data) {
...
+      typeStr = typeStr.slice(3);
+      if (typeStr.endsWith('[]')) {
+        typeStr = typeStr.slice(0, -2);
       }
-      if (param.rest) {
-        type = `...${type}[]`;
+    }
+
+    // Parse union type strings
+    const parseUnion = unionStr => {
```
Comment (davepagurek) [inline utils/convert.mjs:584]: Rather than parsing out arrays, tuples, etc here, could we modify `typeObject` to return something structured? Since we start with the structure, rather than condensing it and re-parsing it, we could just keep the original structure.

**Review by davepagurek — COMMENTED** [review]
Thanks for your work on this, it's looking good!

So it looks like currently we're using `typeObject` and then have to re-parse structure out of the strings it returns. I'm wondering if maybe we can make a new function, e.g. `structuredTypeObject`, that does something similar to the original but preserves more of the original structure instead of condensing it into a string, while keeping around the usual `typeObject` function so that the output `data.json` (which is used on the p5 website) doesn't change?

**Comment by IIITM-Jay** [comment]
Hi @davepagurek Thank you so much for reviewing.
> so that the output data.json (which is used on the p5 website) doesn't change?

Hi @davepagurek for your first concerns above, have you checked this PR by running in local, as in my opinion nothing got changed in the output of `data.json` , there is only a block got added for `generateZodSchemasForFunc` and that too because I have added the below comments and  in the `param_validator.js`, I wonder if you have checked this file for your worries of getting `data.json` output format changed, will remove that if you wish, but it has nothing to do with the changes done in `convert.mjs`. Perhaps, seems like you have missed `param_validator.js` to look upon :slightly_smiling_face:. Used similar kind of  `JSDoc` for `buildParamDocs()` also in `convert.mjs`. But I don't believe that output of `data.json` got changed. Could you please check once

```js
/**
   * Recursively builds a Zod schema from a type descriptor object.
   *
   * @param {object} typeObj - An object describing the type.
   * @returns {ZodType} - A Zod schema representing the type.
   */
  const buildSchema = typeObj => {
```

File: utils/convert.mjs [hunk 2]
```diff
@@ -512,51 +545,166 @@ function cleanUpClassItems(data) {
...
+
+    // Handle function parameters
+    if (param.type === 'Function') {
+      result.type = 'Function';
+
+      if (param.parameters) {
+        // Convert numeric-keyed object or array to array of processed params
+        if (Array.isArray(param.parameters)) {
+          result.parameters = param.parameters.map(processParam);
+        } else if (typeof param.parameters === 'object') {
```
[the remark this answers is not in the document]
Comment (IIITM-Jay) [reply utils/convert.mjs:563]: yes, it somehow generates as 

```js
{ '0': { type: 'Event' } }
{ '0': { type: 'p5.Geometry' } }
```

File: utils/convert.mjs [hunk 3]
```diff
@@ -231,24 +241,55 @@ function getParams(entry) {
...
+      const key = path[i];
+
+      if (i === path.length - 1) {
+        // Leaf node: assign the type info directly
+        current[key] = { ...typeInfo };
+
+        // If the param is a function, populate parameters explicitly (only if not already done)
+        if (
+          typeInfo.type === 'Function' &&
+          !current[key].parameters &&
```
[the remark this answers is not in the document]
Comment (IIITM-Jay) [reply utils/convert.mjs:266]: Good catch — you're right to raise this. I have already did this in `typeobject` with node type as `FunctionType`:

```js
else if (node.type === 'FunctionType') {
  const parameters = node.params.map(p => typeObject(p));
  const returnType = node.result ? typeObject(node.result).type : undefined;
  return {
    type: 'Function',
    parameters,
    ...(returnType && { returnType })
  };
}
```
This check here in `convert.mjs` was meant as a safeguard in case some tag didn’t pass through `typeObject` properly or got overridden with incomplete type info — for example, if a parent param was documented vaguely or multiple `@param`(s) with overlapping paths were present. That said, if we trust that `typeObject` always runs and always returns the correct structure, then I agree this special-case logic may be redundant and we could enforce that structure more cleanly within `typeObject` itself.

Happy to refactor accordingly if we want to centralize the responsibility in `typeObject`.

**Comment by davepagurek** [comment]
> Just checking to confirm: is the goal here to simplify the usage of `typeObject()` by making it the single source of truth for all type-related processing, so downstream methods using it don’t need to do any further conditional handling?

I think so, since it would simplify the process of  structured --> string --> parsed string --> structured to just structured --> different structured.

The one caveat is that this is just for producing `parameterData.json` for now -- I think it's OK to keep the format the same as it currently is for `data.json` and `data.min.json`, which is also generated from `convert.mjs` and is what gets displayed on the website. For FES, we need the deep structure because we need to check all the properties of objects, whereas on the website, we need a single string for each parameter to display. So I think that maybe means having two functions to produce types, one used when generating each file? One generating structured data, one just generating type strings as before?

**Comment by IIITM-Jay** [comment]
> I think it's OK to keep the format the same as it currently is for data.json and data.min.json, which is also generated from convert.mjs and is what gets displayed on the website.

Yes, I agree — it's important to keep the `data.json` and `data.min.json` formats unchanged, especially since they're used for the website and are derived from `convert.mjs`. That’s exactly why I’ve been careful to make changes in a way that doesn't affect those files.

> So I think that maybe means having two functions ...

Regarding the idea of having two separate functions: I was thinking we could potentially keep things simpler by using a single function that conditionally outputs either the structured type (for `FES`) or the stringified type (for website display), depending on the use case. I’ll explore both approaches — separate vs conditional within one — and choose the one that offers the best balance of clarity, maintainability, and long-term flexibility, as to keep as simpler as possible for the longer run sustainability.

**Comment by davepagurek** [comment]
Sounds good, thanks @IIITM-Jay!

**Comment by limzykenneth** [comment]
Just want to make sure that we are keeping things compatible and not duplicating things between this PR and #7863 (I don't think it is but just to mention), since they are touching related things in parallel.

One thing I think would be very nice to have for this part of FES, or more specifically the parameter checks part of FES, is to have some unit tests that includes the edge cases identified, while we continue to refine this part so that we don't have to worry about regression. If it is simple to do at least a skeleton of it, we can do it in this PR or if it looks like a larger endavour we can do it separately and get this in first.

**Comment by davepagurek** [comment]
Also, just wanted to check in again @IIITM-Jay, I think the last things to do here were to update `typeObject` to remove the need to re-parse any strings, and to confirm again afterwards that `data.json` is unmodified. Let me know if there's anything else I can help with to get this across the finish line and merged 🙂

**Comment by ksen0** [comment]
Hi @IIITM-Jay, thanks so much for all your work on this. Could you confirm please if you are still working on this PR? Please check in in the next 7 days (no need to be fully finished in the next 7 days, but to check in whether you are working on it still), otherwise we will finish up this PR and get it merged in. Thank you!
